### PR TITLE
Add VM Support for Marketplace Purchase Plans 

### PR DESCRIPTION
**Pull request by patrickdillon**
VMs created from images in the marketplace or shared galleries may require purchase plans. This adds logic to pass through specified plan data when appropriate.

Simplified version of this upstream logic: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d2b4e8ed2313adc270a0911d9d0c64e344f34cc4/azure/services/virtualmachines/virtualmachines.go#L259-L287

/hold
Depends on openshift/api#1057

### CODE REVIEW

File: pkg/cloud/azure/services/virtualmachines/virtualmachines.go [hunk 1]
```diff
@@ -385,3 +386,21 @@ func getSpotVMOptions(spotVMOptions *machinev1.SpotVMOptions) (compute.VirtualMa
...
+
+func generateImagePlan(image machinev1.Image) *compute.Plan {
+	// If it's not a third-party image or from the shared gallery,
+	// we don't need a purchase plan.
+	if !image.ThirdPartyImage && image.SharedGallery == nil {
+		return nil
+	}
+
+	if image.Publisher == "" || image.SKU == "" || image.Offer == "" {
+		return nil
```
Comment (JoelSpeed) [inline pkg/cloud/azure/services/virtualmachines/virtualmachines.go:398]: Do we care about the version? IIRC the upstream implementation cared about versions and put those into the plan? I could be mistaken though

File: pkg/cloud/azure/services/virtualmachines/virtualmachines_test.go [hunk 2]
```diff
@@ -92,6 +92,23 @@ func TestDeriveVirtualMachineParameters(t *testing.T) {
...
+		{
+			name: "ThirdParty Marketplace Image",
+			updateSpec: func(vmSpec *Spec) {
+				vmSpec.Image.ThirdPartyImage = true
+			},
+			validate: func(g *WithT, vm *compute.VirtualMachine) {
+				g.Expect(vm.Plan).ToNot(BeNil())
+
+			},
+		},
```
Comment (JoelSpeed) [inline pkg/cloud/azure/services/virtualmachines/virtualmachines_test.go:111]: Should we add a ShareImageGallery test as well?

Comment (patrickdillon) [reply pkg/cloud/azure/services/virtualmachines/virtualmachines_test.go:111]: Just removed support for Image Galleries from the API. Apparently we are only doing marketplace images.

File: pkg/cloud/azure/services/virtualmachines/virtualmachines.go (hunk 1 above)
Comment (patrickdillon) [reply pkg/cloud/azure/services/virtualmachines/virtualmachines.go:398]: We don't care about version in the plan. 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d2b4e8ed2313adc270a0911d9d0c64e344f34cc4/azure/services/virtualmachines/virtualmachines.go#L278

**Comment by patrickdillon** [comment]
/hold cancel 
openshift/api#1057 merged

**Review by alexander-demichev — APPROVED** [review]
/lgtm

**Pull request merged**